Error: panic: commit is not available

When trying out git-sizer on an internal repo I am getting following error:

```
Processing blobs: 424652
Processing trees: 407283
panic: commit is not available

goroutine 1 [running]:
github.com/github/git-sizer/sizes.(*Graph).GetCommitSize(0xc420001380, 0xc37d1f0ae042c3b1, 0x181a181366861aad, 0xd4440d70, 0xb300000010)
	/home/mhagger/github/proj/git-sizer/git/.gopath/src/github.com/github/git-sizer/sizes/graph.go:630 +0xc6
github.com/github/git-sizer/sizes.(*Graph).RegisterCommit(0xc420001380, 0x775ac44301b3a032, 0x39b7ad7765e2bb53, 0xc4e6d0cfba, 0xc421f50180)
	/home/mhagger/github/proj/git-sizer/git/.gopath/src/github.com/github/git-sizer/sizes/graph.go:656 +0x1e1
github.com/github/git-sizer/sizes.ScanRepositoryUsingGraph(0xc42000e5d0, 0x116c6f0, 0x2, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/mhagger/github/proj/git-sizer/git/.gopath/src/github.com/github/git-sizer/sizes/graph.go:281 +0x11e8
main.mainImplementation(0x0, 0x0)
	/home/mhagger/github/proj/git-sizer/git/.gopath/src/github.com/github/git-sizer/git-sizer.go:150 +0x8ef
main.main()
	/home/mhagger/github/proj/git-sizer/git/.gopath/src/github.com/github/git-sizer/git-sizer.go:46 +0x26
```

Oops. I am sorry, of course `git fsck` doesn't fix anything.

It doesn't report any errors on my repo either:
```
$ git fsck --no-dangling
Checking object directories: 100% (256/256), done.
Checking objects: 100% (2265686/2265686), done.
Checking connectivity: 1932408, done.

$ ~/bin/git-sizer
Processing blobs: 700857
Processing trees: 998950
panic: commit is not available

goroutine 1 [running]:
github.com/github/git-sizer/sizes.(*Graph).GetCommitSize(0xc4200c6000, 0x765aca5c538afed6, 0xf79c1b412029dceb, 0x5f4071f7, 0xc30000000f)
	/home/mhagger/github/proj/git-sizer/git/.gopath/src/github.com/github/git-sizer/sizes/graph.go:630 +0xc6
github.com/github/git-sizer/sizes.(*Graph).RegisterCommit(0xc4200c6000, 0x1efdd41421eb6764, 0x9c655f3f1847d6b, 0xc4cef55c9a, 0xc43b2ec880)
	/home/mhagger/github/proj/git-sizer/git/.gopath/src/github.com/github/git-sizer/sizes/graph.go:656 +0x1e1
github.com/github/git-sizer/sizes.ScanRepositoryUsingGraph(0xc4200c2000, 0x5687e8, 0x2, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/mhagger/github/proj/git-sizer/git/.gopath/src/github.com/github/git-sizer/sizes/graph.go:281 +0x11e8
main.mainImplementation(0x0, 0x0)
	/home/mhagger/github/proj/git-sizer/git/.gopath/src/github.com/github/git-sizer/git-sizer.go:150 +0x8ef
main.main()
	/home/mhagger/github/proj/git-sizer/git/.gopath/src/github.com/github/git-sizer/git-sizer.go:46 +0x26
```

--- a/README.md
+++ b/README.md
@@ -78,7 +78,7 @@ Is your Git repository bursting at the seams?
 
     b. Build and install from source. See the instructions in [`docs/BUILDING.md`](docs/BUILDING.md).
 
-3.  Change to the directory containing the Git repository that you'd like to analyze, then run
+3.  Change to the directory containing a full, non-shallow clone of the Git repository that you'd like to analyze. Then run
 
         git sizer [<option>...]
 

--- a/git/git.go
+++ b/git/git.go
@@ -64,6 +64,17 @@ type Repository struct {
 	path string
 }
 
+// smartJoin returns the path that can be described as `relPath`
+// relative to `path`, given that `path` is either absolute or is
+// relative to the current directory.
+func smartJoin(path, relPath string) string {
+	if filepath.IsAbs(relPath) {
+		return relPath
+	} else {
+		return filepath.Join(path, relPath)
+	}
+}
+
 func NewRepository(path string) (*Repository, error) {
 	cmd := exec.Command("git", "-C", path, "rev-parse", "--git-dir")
 	out, err := cmd.Output()
@@ -87,10 +98,7 @@ func NewRepository(path string) (*Repository, error) {
 			return nil, err
 		}
 	}
-	gitDir := string(bytes.TrimSpace(out))
-	if !filepath.IsAbs(gitDir) {
-		gitDir = filepath.Join(path, gitDir)
-	}
+	gitDir := smartJoin(path, string(bytes.TrimSpace(out)))
 	repo := &Repository{
 		path: gitDir,
 	}

--- a/git/git.go
+++ b/git/git.go
@@ -99,10 +99,24 @@ func NewRepository(path string) (*Repository, error) {
 		}
 	}
 	gitDir := smartJoin(path, string(bytes.TrimSpace(out)))
-	repo := &Repository{
-		path: gitDir,
+
+	cmd = exec.Command("git", "rev-parse", "--git-path", "shallow")
+	cmd.Dir = gitDir
+	out, err = cmd.Output()
+	if err != nil {
+		return nil, errors.New(
+			fmt.Sprintf(
+				"could not run 'git rev-parse --git-path shallow': %s", err,
+			),
+		)
 	}
-	return repo, nil
+	shallow := smartJoin(gitDir, string(bytes.TrimSpace(out)))
+	_, err = os.Lstat(shallow)
+	if err == nil {
+		return nil, errors.New("this appears to be a shallow clone; full clone required")
+	}
+
+	return &Repository{path: gitDir}, nil
 }
 
 func (repo *Repository) gitCommand(callerArgs ...string) *exec.Cmd {

--- a/git/git.go
+++ b/git/git.go
@@ -97,9 +97,21 @@ func NewRepository(path string) (*Repository, error) {
 	return repo, nil
 }
 
-func (repo *Repository) gitCommand(args ...string) *exec.Cmd {
+func (repo *Repository) gitCommand(callerArgs ...string) *exec.Cmd {
+	// Disable replace references when running our commands:
+	args := []string{"--no-replace-objects"}
+
+	args = append(args, callerArgs...)
+
 	cmd := exec.Command("git", args...)
-	cmd.Env = append(os.Environ(), "GIT_DIR="+repo.path)
+
+	cmd.Env = append(
+		os.Environ(),
+		"GIT_DIR="+repo.path,
+		// Disable grafts when running our commands:
+		"GIT_GRAFT_FILE="+os.DevNull,
+	)
+
 	return cmd
 }
 

